driver/pace/driver/grid.py

```diff
@@ -0,0 +1,273 @@
+import abc
+import dataclasses
+import logging
+import os
+from typing import ClassVar, Optional, Tuple
+
+import f90nml
+import xarray as xr
+
+import pace.driver
+import pace.dsl
+import pace.physics
+import pace.stencils
+import pace.util.grid
+from pace.stencils.testing import TranslateGrid
+from pace.util import CubedSphereCommunicator, QuantityFactory
+from pace.util.grid import (
+    DampingCoefficients,
+    DriverGridData,
+    GridData,
+    MetricTerms,
+    direct_transform,
+)
+from pace.util.namelist import Namelist
+
+from .registry import Registry
+
+
+logger = logging.getLogger(__name__)
+
+
+class GridInitializer(abc.ABC):
+    @abc.abstractmethod
+    def get_grid(
+        self,
+        quantity_factory: pace.util.QuantityFactory,
+        communicator: pace.util.CubedSphereCommunicator,
+    ) -> Tuple[DampingCoefficients, DriverGridData, GridData]:
+        ...
+
+
+@dataclasses.dataclass
+class GridInitializerSelector(GridInitializer):
+    """
+    Dataclass for selecting the implementation of GridInitializer to use.
+
+    Used to circumvent the issue that dacite expects static class definitions,
+    but we would like to dynamically define which Initializer to use. Does this
+    by representing the part of the yaml specification that asks which initializer
+    to use, but deferring to the implementation in that initializer when called.
+    """
+
+    type: str
+    config: GridInitializer
+    registry: ClassVar[Registry] = Registry()
+
+    @classmethod
+    def register(cls, type_name):
+        return cls.registry.register(type_name)
+
+    def get_grid(
+        self,
+        quantity_factory: QuantityFactory,
+        communicator: CubedSphereCommunicator,
+    ) -> Tuple[DampingCoefficients, DriverGridData, GridData]:
+        return self.config.get_grid(
+            quantity_factory=quantity_factory, communicator=communicator
+        )
+
+    @classmethod
+    def from_dict(cls, config: dict):
+        instance = cls.registry.from_dict(config)
+        return cls(config=instance, type=config["type"])
+
+
+@GridInitializerSelector.register("generated")
+@dataclasses.dataclass
+class GeneratedConfig(GridInitializer):
+    """
+    Configuration for baroclinic initialization.
+
+    Attributes:
+        stretch_grid: whether to Schmidt transform the grid
+            (local refinement)
+        stretch_factor: refinement amount
+        lon_target: desired center longitude for refined tile (deg)
+        lat_target: desired center latitude for refined tile (deg)
+        tc_ks: something to do with friction in the vertical ???Ajda
+        restart_path: path to restart data
+    """
+
+    stretch_grid: bool = False
+    stretch_factor: Optional[float] = None
+    lon_target: Optional[float] = None
+    lat_target: Optional[float] = None
+    ks: int = 0
```

Comment:
probably ignore me: It would be cool if stretch_grid wasn't an option, but instead stretch_factor is by default 1.0 (or 0.0, idk which but no stretching) and lon/lat_target are by default the south pole, so that by default no stretching is done but there isn't an option to "turn on" the stretching.